core/src/main/scala/org/http4s/multipart/MultipartParser.scala

```diff
@@ -17,47 +17,67 @@
 package org.http4s
 package multipart
 
-import cats._
 import cats.effect.Concurrent
 import cats.syntax.all._
 import fs2.{Chunk, Pipe, Pull, Pure, Stream}
 import fs2.io.file.Files
 import java.nio.file.{Path, StandardOpenOption}
 import org.typelevel.ci.CIString
+import fs2.RaiseThrowable
 
 /** A low-level multipart-parsing pipe.  Most end users will prefer EntityDecoder[Multipart]. */
 object MultipartParser {
-  private[this] val logger = org.log4s.getLogger
-
   private[this] val CRLFBytesN = Array[Byte]('\r', '\n')
   private[this] val DoubleCRLFBytesN = Array[Byte]('\r', '\n', '\r', '\n')
   private[this] val DashDashBytesN = Array[Byte]('-', '-')
   private[this] val BoundaryBytesN: Boundary => Array[Byte] = boundary =>
     boundary.value.getBytes("UTF-8")
   val StartLineBytesN: Boundary => Array[Byte] = BoundaryBytesN.andThen(DashDashBytesN ++ _)
 
+  /** `delimiter` in RFC 2046 */
   private[this] val ExpectedBytesN: Boundary => Array[Byte] =
     BoundaryBytesN.andThen(CRLFBytesN ++ DashDashBytesN ++ _)
   private[this] val dashByte: Byte = '-'.toByte
   private[this] val streamEmpty = Stream.empty
-  private[this] val PullUnit = Pull.pure[Pure, Unit](())
 
   private type SplitStream[F[_]] = Pull[F, Nothing, (Stream[F, Byte], Stream[F, Byte])]
-  private type SplitFileStream[F[_]] =
-    Pull[F, Nothing, (Stream[F, Byte], Stream[F, Byte], Option[Path])]
+
+  private[this] sealed trait Event
+  private[this] final case class PartStart(value: Headers) extends Event
+  private[this] final case class PartChunk(value: Chunk[Byte]) extends Event
+  private[this] final case object PartEnd extends Event
 
   def parseStreamed[F[_]: Concurrent](
       boundary: Boundary,
       limit: Int = 1024): Pipe[F, Byte, Multipart[F]] = { st =>
-    ignorePrelude[F](boundary, st, limit)
-      .fold(Vector.empty[Part[F]])(_ :+ _)
+    st.through(
+      parseToPartsStream(boundary, limit)
+    ).fold(Vector.empty[Part[F]])(_ :+ _)
       .map(Multipart(_, boundary))
   }
 
   def parseToPartsStream[F[_]: Concurrent](
       boundary: Boundary,
       limit: Int = 1024): Pipe[F, Byte, Part[F]] = { st =>
-    ignorePrelude[F](boundary, st, limit)
+    st.through(
+      parseEvents[F](boundary, limit)
+    )
+      // The left half is the part under construction, the right half is a part to be emitted.
+      .mapAccumulate[Option[Part[F]], Option[Part[F]]](None) { (acc, item) =>
+        (acc, item) match {
+          case (None, PartStart(headers)) =>
+            (Some(Part(headers, Stream.empty)), None)
+          case (Some(acc0), PartChunk(chunk)) =>
+            (Some(acc0.copy(body = acc0.body ++ Stream.chunk(chunk))), None)
+          case (Some(_), PartEnd) =>
+            // Part done - emit it and start over.
+            (None, acc)
+          case _ =>
+            // Shouldn't happen if the `parseToEventsStream` contract holds.
+            sys.error("Unexpected state")
```

Comment:
We have an internal.bug function that refers to our issue tracker for things like this.

Reply:
I've fleshed out these assertion errors a bit. They now go through the effect, and I've written out the actual situations to check exhaustivity. Not that it would help much - AFAICT, exhaustivity warnings don't fail the build here:
```
[warn] /home/nigredo/dev/http4s/core/src/main/scala/org/http4s/multipart/MultipartParser.scala:67:9: match may not be exhaustive.
[warn] It would fail on the following input: (Some(_), PartStart(_))
[warn]         (acc, item) match {
[warn]         ^
```